flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java

```diff
@@ -0,0 +1,481 @@
+
+	private volatile boolean hasAssignedShards;
+
+	private volatile boolean running = true;
+
+
+	// ------------------------------------------------------------------------
+	//  Constructors
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Creates a new Flink Kinesis Consumer.
+	 *
+	 * <p>The AWS credentials to be used, AWS region of the Kinesis streams, initial position to start streaming
+	 * from are configured with a {@link Properties} instance.</p>
+	 *
+	 * @param stream
+	 *           The single AWS Kinesis stream to read from.
+	 * @param deserializer
+	 *           The deserializer used to convert raw bytes of Kinesis records to Java objects (without key).
+	 * @param configProps
+	 *           The properties used to configure AWS credentials, AWS region, and initial starting position.
+	 */
+	public FlinkKinesisConsumer(String stream, DeserializationSchema<T> deserializer, Properties configProps) {
+		this(stream, new KinesisDeserializationSchemaWrapper<T>(deserializer), configProps);
+	}
+
+	/**
+	 * Creates a new Flink Kinesis Consumer.
+	 *
+	 * <p>The AWS credentials to be used, AWS region of the Kinesis streams, initial position to start streaming
+	 * from are configured with a {@link Properties} instance.</p>
+	 *
+	 * @param stream
+	 *           The single AWS Kinesis stream to read from.
+	 * @param deserializer
+	 *           The keyed deserializer used to convert raw bytes of Kinesis records to Java objects.
+	 * @param configProps
+	 *           The properties used to configure AWS credentials, AWS region, and initial starting position.
+	 */
+	public FlinkKinesisConsumer(String stream, KinesisDeserializationSchema<T> deserializer, Properties configProps) {
+		this(Collections.singletonList(stream), deserializer, configProps);
+	}
+
+	/**
+	 * Creates a new Flink Kinesis Consumer.
+	 *
+	 * <p>The AWS credentials to be used, AWS region of the Kinesis streams, initial position to start streaming
+	 * from are configured with a {@link Properties} instance.</p>
+	 *
+	 * @param streams
+	 *           The AWS Kinesis streams to read from.
+	 * @param deserializer
+	 *           The keyed deserializer used to convert raw bytes of Kinesis records to Java objects.
+	 * @param configProps
+	 *           The properties used to configure AWS credentials, AWS region, and initial starting position.
+	 */
+	public FlinkKinesisConsumer(List<String> streams, KinesisDeserializationSchema<T> deserializer, Properties configProps) {
+		checkNotNull(streams, "streams can not be null");
+
+		this.configProps = checkNotNull(configProps, "configProps can not be null");
+
+		// check the configuration properties for any conflicting settings
+		validatePropertiesConfig(this.configProps);
+
+		this.deserializer = checkNotNull(deserializer, "deserializer can not be null");
+
+		this.shards = new KinesisProxy(configProps).getShardList(streams);
```

Comment:
I'm currently trying to mock the KinesisProxy here for unit tests.
Currently considering two ways of modifying this part to make the class testable:

1) Introduce a serializable concrete `KinesisProxyFactory` class which will be instantiated as a default field within `FlinkKinesisConsumer`. The `KinesisProxyFactory` will have a non-static method to help create the proxy. This line will this be changed to,
`this.shards = this.kinesisProxyFactory.createProxy(configProps).getShardList(streams)` which is much easier to mock.

2) Change this line to,
`this.shards = createProxy(configProps).getShardList(streams)`, where `createProxy` is a protected class method within `FlinkKinesisConsumer` that does the actual instantiation of `KinesisProxy`. In our tests, we can write a `TestableFlinkKinesisConsumer` that extends `FlinkKinesisConsumer` and overrides the protected `createProxy` to return a mocked `KinesisProxy`.

Downside of approach 1 is that `FlinkKinesisConsumer` will be carrying an extra `KinesisProxyFactory` field instance. On the other hand, approach 2 is also quite tedious because `TestableFlinkKinesisConsumer` will need to also implement dummies for all variants of `FlinkKinesisConsumer`'s constructors ...

I'm not quite familiar with writing good tests, so any suggestion here will be quite helpful!